### PR TITLE
Add zipformer recipe for audio tagging

### PR DESCRIPTION
**Pull request by marcoyang1998**
This PR adds a recipe for audio tagging using Zipformer. We use AudioSet as the training set.

Without using knowledge distillation, the Zipformer-M model achieves mAP value of 45.1. Using knowledge distillation could further improve the results (will add support for this in another PR)

- [x] Add usage example
- [x] Upload the pre-trained model

### CODE REVIEW

**Comment by marcoyang1998** [comment]
You can use the pretrained model to predict the audio events of an audio clip through the following command:

```bash
python zipformer/pretrained.py \
    --checkpoint zipformer/exp_at_as_full/pretrained.pt \
    --label-dict downloads/audioset/class_labels_indices.csv \
    downloads/audioset/eval/wav/__p-iA312kg_70.000_80.000.wav
```

You should see the following output:
```
2024-03-20 18:40:35,611 INFO [pretrained.py:175] Reading sound files: ['downloads/audioset/eval/wav/__p-iA312kg_70.000_80.000.wav']
2024-03-20 18:40:35,616 INFO [pretrained.py:181] Decoding started
Top 5 predicted labels of the 0 th audio are ['Cat', 'Animal', 'Domestic animals, pets', 'Meow', 'Caterwaul'] with probability of [0.9497291445732117, 0.8830551505088806, 0.8787704110145569, 0.2819702923297882, 0.18947316706180573]
```

**Comment by marcoyang1998** [comment]
The pretrained model can be found here: https://huggingface.co/marcoyang/icefall-audio-tagging-audioset-zipformer-2024-03-12#/

File: egs/audioset/AT/local/generate_audioset_manifest.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+import torch
+import torchaudio
+import logging
+import glob
+from lhotse import load_manifest, CutSet, Fbank, FbankConfig, LilcomChunkyWriter
+from lhotse.cut import MonoCut
+from lhotse.audio import Recording
+from lhotse.supervision import SupervisionSegment
+from argparse import ArgumentParser
```
Comment (csukuangfj) [inline egs/audioset/AT/local/generate_audioset_manifest.py:12]: please sort the import with `isort`, which follows https://peps.python.org/pep-0008/#imports

```
Imports should be grouped in the following order:

1. Standard library imports.
2. Related third party imports.
3. Local application/library specific imports.
```

Comment (marcoyang1998) [reply egs/audioset/AT/local/generate_audioset_manifest.py]: Fixed

File: egs/audioset/AT/local/generate_audioset_manifest.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+from lhotse import CutSet, Fbank, FbankConfig, LilcomChunkyWriter
+from lhotse.audio import Recording
+from lhotse.cut import MonoCut
+from lhotse.supervision import SupervisionSegment
+
+from icefall.utils import get_executor
+
+torch.set_num_threads(1)
+torch.set_num_interop_threads(1)
+
```
Comment (csukuangfj) [inline egs/audioset/AT/local/generate_audioset_manifest.py:16]: Could you follow
https://github.com/k2-fsa/icefall/blob/42de4591107868816d240c991b1e8acdfdfd090b/egs/librispeech/ASR/local/compute_fbank_librispeech.py#L19-L24
to give a short description about what this file does?

File: egs/audioset/AT/local/generate_audioset_manifest.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+def parse_csv(csv_file="downloads/audioset/full_train_asedata_with_duration.csv"):
+
```
Comment (csukuangfj) [inline egs/audioset/AT/local/generate_audioset_manifest.py:19]: Could you give a sample line in the comment?
It helps read the code if a sample line is available.

File: egs/audioset/AT/local/generate_audioset_manifest.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    dataset_dir = args.dataset_dir
+    split = args.split
+    feat_output_dir = args.feat_output_dir
+
+    num_jobs = 15
+    num_mel_bins = 80
+
+    import pdb
+
+    pdb.set_trace()
```
Comment (csukuangfj) [inline egs/audioset/AT/local/generate_audioset_manifest.py:67]:
```suggestion
```

Please remove the debug statements.

File: egs/audioset/AT/local/generate_audioset_manifest.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+def main():
+    parser = get_parser()
+    args = parser.parse_args()
+
+    dataset_dir = args.dataset_dir
+    split = args.split
+    feat_output_dir = args.feat_output_dir
+
+    num_jobs = 15
```
Comment (csukuangfj) [inline egs/audioset/AT/local/generate_audioset_manifest.py:62]: Please follow other scripts to set the `num_jobs`.

Users may not have 15 CPUs on their machines.

Please see also
https://github.com/k2-fsa/icefall/blob/42de4591107868816d240c991b1e8acdfdfd090b/egs/librispeech/ASR/local/compute_fbank_librispeech.py#L81

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/audioset/AT/local/generate_audioset_manifest.py (hunk 3 above)
Comment (csukuangfj) [inline egs/audioset/AT/local/generate_audioset_manifest.py:19]: Please delete this empty line.

File: egs/audioset/AT/local/generate_audioset_manifest.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+from lhotse.cut import MonoCut
+from lhotse.supervision import SupervisionSegment
+
+from icefall.utils import get_executor
+
+torch.set_num_threads(1)
+torch.set_num_interop_threads(1)
+
+
+def parse_csv(csv_file="downloads/audioset/full_train_asedata_with_duration.csv"):
```
Comment (csukuangfj) [inline egs/audioset/AT/local/generate_audioset_manifest.py:18]: Looks like the default value is never used. Could you remove it?

File: egs/audioset/AT/local/generate_audioset_manifest.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            recording_id=cut.recording.id,
+            start=0.0,
+            channel=0,
+            duration=cut.duration,
+        )
+        try:
+            supervision.audio_event = labels[cut_id]
+        except KeyError:
+            logging.info(f"No labels found for {cut_id}.")
+            supervision.audio_event = ""
```
Comment (csukuangfj) [inline egs/audioset/AT/local/generate_audioset_manifest.py:103]: Can we drop such cuts if its `audio_event` is not available?

File: egs/audioset/AT/local/generate_audioset_manifest.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+    cuts = CutSet.from_cuts(new_cuts)
+
+    extractor = Fbank(FbankConfig(num_mel_bins=num_mel_bins))
+
+    logging.info(f"Computing fbank features for {split}")
+    with get_executor() as ex:
+        cuts = cuts.compute_and_store_features(
+            extractor=extractor,
+            storage_path=f"{feat_output_dir}/{split}_{args.split}_feats",
```
Comment (csukuangfj) [inline egs/audioset/AT/local/generate_audioset_manifest.py:118]:
```suggestion
            storage_path=f"{feat_output_dir}/{split}_{split}_feats",
```

`split` is shorter than `args.split`.

File: egs/audioset/AT/zipformer/at_datamodule.py [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+    def __call__(self, worker_id: int):
+        fix_random_seed(self.seed + worker_id)
+
+
+class AudioSetATDatamodule:
+    """
+    DataModule for k2 audio tagging (AT) experiments.
+
+
+    It contains all the common data pipeline modules used in ASR
```
Comment (csukuangfj) [inline egs/audioset/AT/zipformer/at_datamodule.py:59]:
```suggestion
    It contains all the common data pipeline modules used in AT
```

File: egs/audioset/AT/zipformer/at_datamodule.py [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+    experiments, e.g.:
+    - dynamic batch size,
+    - bucketing samplers,
+    - cut concatenation,
+    - augmentation,
+    - on-the-fly feature extraction
+
+    This class should be derived for specific corpora used in ASR tasks.
```
Comment (csukuangfj) [inline egs/audioset/AT/zipformer/at_datamodule.py:67]:
```suggestion
    This class should be derived for specific corpora used in AT tasks.
```

File: egs/audioset/AT/local/generate_audioset_manifest.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
+import argparse
```
Comment (csukuangfj) [inline egs/audioset/AT/local/generate_audioset_manifest.py:1]: Could you add a README.md and a RESULTS.md for this recipe?

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/audioset/AT/local/generate_audioset_manifest.py (hunk 11 above)
Comment (csukuangfj) [reply egs/audioset/AT/local/generate_audioset_manifest.py:1]: By the way, it lacks a file `prepare.sh`.

File: egs/audioset/AT/local/generate_audioset_manifest.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        "--split",
+        type=str,
+        default="balanced",
+        choices=["balanced", "unbalanced", "eval", "eval_all"],
+    )
+
+    parser.add_argument(
+        "--feat-output-dir",
+        type=str,
+        default="data/fbank_audioset",
```
Comment (csukuangfj) [inline egs/audioset/AT/local/generate_audioset_manifest.py:48]: https://github.com/k2-fsa/icefall/blob/7a8c9b7f53b35d2b72d2f64f6547c1c367118462/egs/audioset/AT/zipformer/at_datamodule.py#L93

Could you use a consistent default value?

File: egs/audioset/AT/zipformer/at_datamodule.py [hunk 13]
```diff
@@ -0,0 +1,421 @@
...
+            type=str,
+            default="balanced",
+            choices=["balanced", "full"],
+        )
+
+        group.add_argument(
+            "--manifest-dir",
+            type=Path,
+            default=Path("data/fbank"),
+            help="Path to directory with train/valid/test cuts.",
```
Comment (csukuangfj) [inline egs/audioset/AT/zipformer/at_datamodule.py:94]: Please update the help info.

File: egs/audioset/AT/zipformer/at_datamodule.py [hunk 14]
```diff
@@ -0,0 +1,421 @@
...
+            test,
+            batch_size=None,
+            sampler=sampler,
+            num_workers=self.args.num_workers,
+        )
+        return test_dl
+
+    @lru_cache()
+    def audioset_train_cuts(self) -> CutSet:
+        logging.info("About to get the audioset cuts for KD.")
```
Comment (csukuangfj) [inline egs/audioset/AT/zipformer/at_datamodule.py:398]:
```suggestion
        logging.info("About to get the audioset train cuts.")
```

There is no `KD`, I think.

File: egs/audioset/AT/zipformer/at_datamodule.py [hunk 15]
```diff
@@ -0,0 +1,421 @@
...
+                weights=[20000, 2000000],
+                stop_early=True,
+            )
+        else:
+            cuts = balanced_cuts
+        return cuts
+
+    @lru_cache()
+    def audioset_eval_cuts(self) -> CutSet:
+        logging.info("About to get test-other cuts")
```
Comment (csukuangfj) [inline egs/audioset/AT/zipformer/at_datamodule.py:418]:
```suggestion
        logging.info("About to get audioset eval cuts")
```

File: egs/audioset/AT/zipformer/train.py [hunk 16]
```diff
@@ -0,0 +1,1189 @@
...
+        "--left-context-frames",
+        type=str,
+        default="64,128,256,-1",
+        help="Maximum left-contexts for causal training, measured in frames which will "
+        "be converted to a number of chunks.  If splitting into chunks, "
+        "chunk left-context frames will be chosen randomly from this list; else not relevant.",
+    )
+
+    parser.add_argument(
+        "--num-events", type=int, default=527, help="Number of sound events"
```
Comment (csukuangfj) [inline egs/audioset/AT/zipformer/train.py:210]: Can we write this default value into a file in `prepare.sh`?
It is to avoid giving an incorrect value

File: egs/audioset/AT/zipformer/train.py [hunk 17]
```diff
@@ -0,0 +1,1189 @@
...
+        values >= 1.0 are fully warmed up and have all modules present.
+    """
+    device = model.device if isinstance(model, DDP) else next(model.parameters()).device
+    feature = batch["inputs"]
+    # at entry, feature is (N, T, C)
+    assert feature.ndim == 3
+    feature = feature.to(device)
+
+    supervisions = batch["supervisions"]
+    events = supervisions["audio_event"]  # the label indices are in CED format
```
Comment (csukuangfj) [inline egs/audioset/AT/zipformer/train.py:652]: Could you give a link about what CED is?
You can put the github repository link here.

File: egs/audioset/AT/local/generate_audioset_manifest.py (hunk 7 above)
Comment (marcoyang1998) [reply egs/audioset/AT/local/generate_audioset_manifest.py:103]: sure, good idea

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/audioset/AT/local/generate_audioset_manifest.py (hunk 2 above)
Comment (marcoyang1998) [reply egs/audioset/AT/local/generate_audioset_manifest.py:16]: Added

File: egs/audioset/AT/local/generate_audioset_manifest.py (hunk 3 above)
Comment (marcoyang1998) [reply egs/audioset/AT/local/generate_audioset_manifest.py:19]: Added

File: egs/audioset/AT/local/generate_audioset_manifest.py (hunk 12 above)
Comment (marcoyang1998) [reply egs/audioset/AT/local/generate_audioset_manifest.py]: Just made them the same default value

File: egs/audioset/AT/zipformer/at_datamodule.py (hunk 14 above)
Comment (marcoyang1998) [reply egs/audioset/AT/zipformer/at_datamodule.py:398]: Yes, just fixed it.

**Comment by marcoyang1998** [comment]
Since there is no official way of downloading AudioSet now, this recipe assumes that you have your own version of AudioSet. The required structure of the dataset will be given in `prepare.sh` (will add in the next commit).

File: egs/audioset/AT/zipformer/train.py [hunk 18]
```diff
@@ -0,0 +1,1189 @@
...
+    info["loss"] = loss.detach().cpu().item()
+    info["audio_tagging_loss"] = audio_tagging_loss.detach().cpu().item()
+
+    return loss, info
+
+
+def str2multihot(events: List[str], n_classes=527, id_mapping=None):
+    # Convert strings separated by semi-colon to multi-hot class labels
+    # input: ["1;2", "2;3"]
+    # output: torch.tensor([[1,1,0], [0,1,1]])
```
Comment (csukuangfj) [inline egs/audioset/AT/zipformer/train.py:688]: Could you specify the `n_classes` in this example?

The output does not look correct to me.

File: egs/audioset/AT/zipformer/train.py [hunk 19]
```diff
@@ -0,0 +1,1189 @@
...
+
+    with torch.set_grad_enabled(is_training):
+        audio_tagging_loss = model(
+            x=feature,
+            x_lens=feature_lens,
+            target=labels,
+        )
+
+        loss = 0.0
+        loss += audio_tagging_loss
```
Comment (csukuangfj) [inline egs/audioset/AT/zipformer/train.py:669]: Is there a reason to use a separate name `loss` for `audio_tagging_loss`?

```suggestion
        loss = audio_tagging_loss
```

File: egs/audioset/AT/zipformer/train.py (hunk 18 above)
Comment (marcoyang1998) [reply egs/audioset/AT/zipformer/train.py:688]: This dummy example is for `n_classes=3`, the class ID should start from 1.

Comment (csukuangfj) [reply egs/audioset/AT/zipformer/train.py:688]: >  the class ID should start from 1.

Is it also starting from 1 for the training data?

If not,  I suggest that you use a consistent starting value. Otherwise, it is confusing.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/audioset/AT/zipformer/train.py (hunk 19 above)
Comment (marcoyang1998) [reply egs/audioset/AT/zipformer/train.py:669]: The code was modified from another recipe with multiple losses. Will fix this.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/audioset/AT/zipformer/export.py [hunk 20]
```diff
@@ -0,0 +1,324 @@
...
+        help="Path to the tokens.txt",
+    )
+
+    parser.add_argument(
+        "--jit",
+        type=str2bool,
+        default=False,
+        help="""True to save a model after applying torch.jit.script.
+        It will generate a file named jit_script.pt.
+        Check ./jit_pretrained.py for how to use it.
```
[the remark this answers is not in the document]
Comment (marcoyang1998) [reply egs/audioset/AT/zipformer/export.py:158]: I just added the file, could you please have a look?

**Comment by marcoyang1998** [comment]
Added `jit_pretrained.py`. The script produces the same results as `pretrained.py`.

File: egs/audioset/AT/local/generate_audioset_manifest.py (hunk 11 above)
Comment (marcoyang1998) [reply egs/audioset/AT/local/generate_audioset_manifest.py:23]: Added

File: egs/audioset/AT/zipformer/train.py (hunk 17 above)
Comment (marcoyang1998) [reply egs/audioset/AT/zipformer/train.py]: Added

**Comment by marcoyang1998** [comment]
@csukuangfj Can we merge this now?

File: egs/audioset/AT/RESULTS.md [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+### zipformer
+See <https://github.com/k2-fsa/icefall/pull/1421> for more details
+
+[zipformer](./zipformer)
+
+You can find a pretrained model, training logs, decoding logs, and decoding results at:
+<https://huggingface.co/marcoyang/icefall-audio-tagging-audioset-zipformer-2024-03-12#/>
+
+The model achieves the following mean averaged precision on AudioSet:
```
Comment (csukuangfj) [inline egs/audioset/AT/RESULTS.md:11]: Could you  place a link to `AudioSet`?

**Review by csukuangfj — APPROVED** [review]
Thanks! Just left a minor comment.

File: egs/audioset/AT/RESULTS.md (hunk 21 above)
Comment (marcoyang1998) [reply egs/audioset/AT/RESULTS.md:11]: Sure, I will add the link in readme.

**Comment by herbiel** [comment]
可以支持finetune吗

**Comment by marcoyang1998** [comment]
支持的，换一下head就行
